From f5cd7d06ab0b5723ad0d83ea6b516b50efa0bb93 Mon Sep 17 00:00:00 2001
From: Yongjun Zhang <yzhang@cloudera.com>
Date: Tue, 5 Jan 2016 10:58:59 -0800
Subject: [PATCH 1051/2848] HADOOP-12604. Exception may be swallowed in
 KMSClientProvider. (Yongjun Zhang)

(cherry picked from commit 28bd138018bea6fc9c3bfb94c7a4143420f02ced)
(cherry picked from commit f372d3faae48c8e3c54edd1da8efebcaa9ab0e7a)

Conflicts:
	hadoop-common-project/hadoop-common/CHANGES.txt

Change-Id: Icdf7532e236b4e7619f6e8c6b887d53f925e63e3
---
 .../hadoop/crypto/key/kms/KMSClientProvider.java   |   12 +++---------
 1 file changed, 3 insertions(+), 9 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/KMSClientProvider.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/KMSClientProvider.java
index 744c32f..61f6de4 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/KMSClientProvider.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/key/kms/KMSClientProvider.java
@@ -26,6 +26,7 @@
 import org.apache.hadoop.crypto.key.KeyProviderFactory;
 import org.apache.hadoop.fs.CommonConfigurationKeysPublic;
 import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.IOUtils;
 import org.apache.hadoop.io.Text;
 import org.apache.hadoop.security.Credentials;
 import org.apache.hadoop.security.ProviderUtils;
@@ -512,7 +513,7 @@ public HttpURLConnection run() throws Exception {
         writeJson(jsonOutput, conn.getOutputStream());
       }
     } catch (IOException ex) {
-      conn.getInputStream().close();
+      IOUtils.closeStream(conn.getInputStream());
       throw ex;
     }
     if ((conn.getResponseCode() == HttpURLConnection.HTTP_FORBIDDEN
@@ -550,15 +551,8 @@ public HttpURLConnection run() throws Exception {
       try {
         is = conn.getInputStream();
         ret = mapper.readValue(is, klass);
-      } catch (IOException ex) {
-        if (is != null) {
-          is.close();
-        }
-        throw ex;
       } finally {
-        if (is != null) {
-          is.close();
-        }
+        IOUtils.closeStream(is);
       }
     }
     return ret;
-- 
1.7.9.5

